Scheduled job to delete non activated users fails

Scheduled job to delete non activated users fails with following exception. 
The job should check whether it is a participant or other users and delete entries in corresponding tables to avoid foreign-key violations 
**Enviornment**
test
**Date reported**
03-07-2018
**Stack trace**
```
managementportal-app_1         | 2018-07-03 01:00:00.113 ERROR 8 --- [pool-5-thread-1] o.s.s.s.TaskUtils$LoggingErrorHandler    : Unexpected error occurred in scheduled task.
managementportal-app_1         | 
managementportal-app_1         | org.springframework.dao.DataIntegrityViolationException: could not execute statement; SQL [n/a]; constraint [fk_subject_user_id]; nested exception is org.hibernate.exception.ConstraintViolationException: could not execute statement
managementportal-app_1         | 	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:278)
managementportal-app_1         | 	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:244)
managementportal-app_1         | 	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:521)
managementportal-app_1         | 	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:761)
managementportal-app_1         | 	at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:730)
managementportal-app_1         | 	at org.springframework.transaction.interceptor.TransactionAspectSupport.commitTransactionAfterReturning(TransactionAspectSupport.java:518)
managementportal-app_1         | 	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:292)
managementportal-app_1         | 	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
managementportal-app_1         | 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
managementportal-app_1         | 	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:673)
managementportal-app_1         | 	at org.radarcns.management.service.UserService$$EnhancerBySpringCGLIB$$1572bbe9.removeNotActivatedUsers(<generated>)
managementportal-app_1         | 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
managementportal-app_1         | 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
managementportal-app_1         | 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
managementportal-app_1         | 	at java.lang.reflect.Method.invoke(Method.java:498)
managementportal-app_1         | 	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:65)
managementportal-app_1         | 	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
managementportal-app_1         | 	at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:81)
managementportal-app_1         | 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
managementportal-app_1         | 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
managementportal-app_1         | 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
managementportal-app_1         | 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
managementportal-app_1         | 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
managementportal-app_1         | 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
managementportal-app_1         | 	at java.lang.Thread.run(Thread.java:748)
managementportal-app_1         | Caused by: org.hibernate.exception.ConstraintViolationException: could not execute statement
managementportal-app_1         | 	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:112)
managementportal-app_1         | 	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:42)
managementportal-app_1         | 	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:111)
managementportal-app_1         | 	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:97)
managementportal-app_1         | 	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:178)
managementportal-app_1         | 	at org.hibernate.engine.jdbc.batch.internal.NonBatchingBatch.addToBatch(NonBatchingBatch.java:45)
managementportal-app_1         | 	at org.hibernate.persister.entity.AbstractEntityPersister.delete(AbstractEntityPersister.java:3325)
managementportal-app_1         | 	at org.hibernate.persister.entity.AbstractEntityPersister.delete(AbstractEntityPersister.java:3562)
managementportal-app_1         | 	at org.hibernate.action.internal.EntityDeleteAction.execute(EntityDeleteAction.java:99)
managementportal-app_1         | 	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:589)
managementportal-app_1         | 	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:463)
managementportal-app_1         | 	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:337)
managementportal-app_1         | 	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)
managementportal-app_1         | 	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1437)
managementportal-app_1         | 	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:493)
managementportal-app_1         | 	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3207)
managementportal-app_1         | 	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2413)
managementportal-app_1         | 	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:473)
managementportal-app_1         | 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:156)
managementportal-app_1         | 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.access$100(JdbcResourceLocalTransactionCoordinatorImpl.java:38)
managementportal-app_1         | 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:231)
managementportal-app_1         | 	at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:68)
managementportal-app_1         | 	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:517)
managementportal-app_1         | 	... 22 common frames omitted
managementportal-app_1         | Caused by: org.postgresql.util.PSQLException: ERROR: update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
managementportal-app_1         |   Detail: Key (id)=(14257) is still referenced from table "subject".
managementportal-app_1         | 	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2455)
managementportal-app_1         | 	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2155)
managementportal-app_1         | 	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:288)
managementportal-app_1         | 	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:430)
managementportal-app_1         | 	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:356)
managementportal-app_1         | 	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:168)
managementportal-app_1         | 	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:135)
managementportal-app_1         | 	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
managementportal-app_1         | 	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
managementportal-app_1         | 	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:175)
managementportal-app_1         | 	... 40 common frames omitted

```

--- a/src/main/java/org/radarcns/management/service/SubjectService.java
+++ b/src/main/java/org/radarcns/management/service/SubjectService.java
@@ -11,6 +11,7 @@ import static org.radarcns.management.web.rest.errors.ErrorConstants.ERR_SUBJECT
 
 import java.net.MalformedURLException;
 import java.net.URL;
+import java.time.Period;
 import java.time.ZonedDateTime;
 import java.util.Arrays;
 import java.util.Collections;
@@ -36,6 +37,7 @@ import org.radarcns.management.repository.AuthorityRepository;
 import org.radarcns.management.repository.RoleRepository;
 import org.radarcns.management.repository.SourceRepository;
 import org.radarcns.management.repository.SubjectRepository;
+import org.radarcns.management.repository.UserRepository;
 import org.radarcns.management.service.dto.MinimalSourceDetailsDTO;
 import org.radarcns.management.service.dto.SubjectDTO;
 import org.radarcns.management.service.dto.UserDTO;
@@ -54,6 +56,7 @@ import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.data.history.Revisions;
+import org.springframework.scheduling.annotation.Scheduled;
 import org.springframework.security.crypto.password.PasswordEncoder;
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
@@ -97,6 +100,8 @@ public class SubjectService {
     @Autowired
     private ManagementPortalProperties managementPortalProperties;
 
+    @Autowired
+    private UserRepository userRepository;
 
     /**
      * Create a new subject.
@@ -489,5 +494,29 @@ public class SubjectService {
                     OAUTH_CLIENT, ERR_NO_VALID_PRIVACY_POLICY_URL_CONFIGURED,
                     params);
         }
+    }
+
+    /**
+     * Not activated users should be automatically deleted after 3 days. <p> This is scheduled to
+     * get fired everyday, at midnight. Preferably we do this scan before
+     * {@link UserService#removeNotActivatedUsers()}, since this will remove the not activated
+     * user tied to the subject as well.</p>
+     */
+    @Scheduled(cron = "0 0 0 * * ?")
+    public void removeNotActivatedSubjects() {
+        log.info("Scheduled scan for expired subject accounts starting now");
+        ZonedDateTime cutoff = ZonedDateTime.now().minus(Period.ofDays(3));
+
+        // first delete non-activated users related to subjects
+        userRepository.findAllByActivated(false).stream()
+                .filter(user -> revisionService.getAuditInfo(user).getCreatedAt().isBefore(cutoff))
+                .map(User::getLogin)
+                .map(subjectRepository::findOneWithEagerBySubjectLogin)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .forEach(subject -> {
+                    log.info("Deleting not activated subject after 3 days: {}", subject);
+                    subjectRepository.delete(subject);
+                });
     }
 }

--- a/src/main/java/org/radarcns/management/service/UserService.java
+++ b/src/main/java/org/radarcns/management/service/UserService.java
@@ -1,17 +1,5 @@
 package org.radarcns.management.service;
 
-import static org.radarcns.management.web.rest.errors.EntityName.USER;
-
-import java.time.Period;
-import java.time.ZonedDateTime;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Objects;
-import java.util.Optional;
-import java.util.Set;
-
 import org.radarcns.auth.authorization.AuthoritiesConstants;
 import org.radarcns.auth.config.Constants;
 import org.radarcns.management.domain.Project;
@@ -20,6 +8,7 @@ import org.radarcns.management.domain.User;
 import org.radarcns.management.repository.AuthorityRepository;
 import org.radarcns.management.repository.ProjectRepository;
 import org.radarcns.management.repository.RoleRepository;
+import org.radarcns.management.repository.SubjectRepository;
 import org.radarcns.management.repository.UserRepository;
 import org.radarcns.management.repository.filters.UserFilter;
 import org.radarcns.management.security.SecurityUtils;
@@ -40,6 +29,18 @@ import org.springframework.scheduling.annotation.Scheduled;
 import org.springframework.security.crypto.password.PasswordEncoder;
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
+
+import java.time.Period;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import static org.radarcns.management.web.rest.errors.EntityName.USER;
 
 /**
  * Service class for managing users.
@@ -73,6 +74,9 @@ public class UserService {
 
     @Autowired
     private RevisionService revisionService;
+
+    @Autowired
+    private SubjectRepository subjectRepository;
 
     /**
      * Activate a user with the given activation key.
@@ -351,10 +355,22 @@ public class UserService {
     public void removeNotActivatedUsers() {
         log.info("Scheduled scan for expired user accounts starting now");
         ZonedDateTime cutoff = ZonedDateTime.now().minus(Period.ofDays(3));
+
+        // first delete non-activated users related to subjects
         userRepository.findAllByActivated(false).stream()
-                .filter(user ->
-                    revisionService.getAuditInfo(user).getCreatedAt().isBefore(cutoff)
-                )
+                .filter(user -> revisionService.getAuditInfo(user).getCreatedAt().isBefore(cutoff))
+                .map(User::getLogin)
+                .map(subjectRepository::findOneWithEagerBySubjectLogin)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .forEach(subject -> {
+                    log.info("Deleting not activated subject after 3 days: {}", subject);
+                    subjectRepository.delete(subject);
+                });
+
+        // now re-query to find non-activated users not related to subjects and delete them
+        userRepository.findAllByActivated(false).stream()
+                .filter(user -> revisionService.getAuditInfo(user).getCreatedAt().isBefore(cutoff))
                 .forEach(user -> {
                     log.info("Deleting not activated user after 3 days: {}", user.getLogin());
                     userRepository.delete(user);

--- a/src/main/java/org/radarcns/management/service/UserService.java
+++ b/src/main/java/org/radarcns/management/service/UserService.java
@@ -8,7 +8,6 @@ import org.radarcns.management.domain.User;
 import org.radarcns.management.repository.AuthorityRepository;
 import org.radarcns.management.repository.ProjectRepository;
 import org.radarcns.management.repository.RoleRepository;
-import org.radarcns.management.repository.SubjectRepository;
 import org.radarcns.management.repository.UserRepository;
 import org.radarcns.management.repository.filters.UserFilter;
 import org.radarcns.management.security.SecurityUtils;
@@ -23,6 +22,7 @@ import org.radarcns.management.web.rest.errors.NotFoundException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DataIntegrityViolationException;
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.scheduling.annotation.Scheduled;
@@ -74,9 +74,6 @@ public class UserService {
 
     @Autowired
     private RevisionService revisionService;
-
-    @Autowired
-    private SubjectRepository subjectRepository;
 
     /**
      * Activate a user with the given activation key.
@@ -356,24 +353,15 @@ public class UserService {
         log.info("Scheduled scan for expired user accounts starting now");
         ZonedDateTime cutoff = ZonedDateTime.now().minus(Period.ofDays(3));
 
-        // first delete non-activated users related to subjects
-        userRepository.findAllByActivated(false).stream()
-                .filter(user -> revisionService.getAuditInfo(user).getCreatedAt().isBefore(cutoff))
-                .map(User::getLogin)
-                .map(subjectRepository::findOneWithEagerBySubjectLogin)
-                .filter(Optional::isPresent)
-                .map(Optional::get)
-                .forEach(subject -> {
-                    log.info("Deleting not activated subject after 3 days: {}", subject);
-                    subjectRepository.delete(subject);
-                });
-
-        // now re-query to find non-activated users not related to subjects and delete them
         userRepository.findAllByActivated(false).stream()
                 .filter(user -> revisionService.getAuditInfo(user).getCreatedAt().isBefore(cutoff))
                 .forEach(user -> {
-                    log.info("Deleting not activated user after 3 days: {}", user.getLogin());
-                    userRepository.delete(user);
+                    try {
+                        userRepository.delete(user);
+                        log.info("Deleted not activated user after 3 days: {}", user.getLogin());
+                    } catch (DataIntegrityViolationException ex) {
+                        log.error("Could not delete user with login " +  user.getLogin(), ex);
+                    }
                 });
     }
 

--- a/src/test/java/org/radarcns/management/service/SubjectServiceTest.java
+++ b/src/test/java/org/radarcns/management/service/SubjectServiceTest.java
@@ -1,23 +1,38 @@
 package org.radarcns.management.service;
 
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
 import static org.radarcns.management.service.dto.ProjectDTO.PRIVACY_POLICY_URL;
 import static org.radarcns.management.service.dto.SubjectDTO.SubjectStatus.ACTIVATED;
+import static org.radarcns.management.web.rest.TestUtil.commitTransactionAndStartNew;
 
 import java.net.URL;
+import java.time.Period;
+import java.time.ZonedDateTime;
 import java.util.Collections;
+import java.util.Date;
+import java.util.List;
 
+import org.hibernate.envers.AuditReader;
+import org.hibernate.envers.query.AuditEntity;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.radarcns.management.ManagementPortalTestApp;
 import org.radarcns.management.domain.Subject;
+import org.radarcns.management.domain.User;
+import org.radarcns.management.domain.audit.CustomRevisionEntity;
+import org.radarcns.management.repository.SubjectRepository;
+import org.radarcns.management.repository.UserRepository;
 import org.radarcns.management.service.dto.ProjectDTO;
 import org.radarcns.management.service.dto.SubjectDTO;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
 import org.springframework.test.context.junit4.SpringRunner;
+import org.springframework.test.util.ReflectionTestUtils;
 import org.springframework.transaction.annotation.Transactional;
+
+import javax.persistence.EntityManager;
 
 @RunWith(SpringRunner.class)
 @SpringBootTest(classes = ManagementPortalTestApp.class)
@@ -47,6 +62,15 @@ public class SubjectServiceTest {
 
     @Autowired
     private ProjectService projectService;
+
+    @Autowired
+    private SubjectRepository subjectRepository;
+
+    @Autowired
+    private RevisionService revisionService;
+
+    @Autowired
+    private UserRepository userRepository;
 
 
     /**
@@ -87,5 +111,45 @@ public class SubjectServiceTest {
         assertNotNull(privacyPolicyUrl);
         assertEquals(privacyPolicyUrl.toExternalForm(), DEFAULT_PROJECT_PRIVACY_POLICY_URL);
 
+    }
+
+    @Test
+    public void testFindNotActivatedSubjectsByCreationDateBefore() {
+        User expiredUser = UserServiceIntTest.addExpiredUser(userRepository);
+        Subject expiredSubject = new Subject();
+        expiredSubject.setUser(expiredUser);
+        subjectRepository.save(expiredSubject);
+        commitTransactionAndStartNew();
+
+        AuditReader auditReader = ((AuditReader) ReflectionTestUtils
+                .getField(revisionService, "auditReader"));
+        Object[] firstRevision = (Object[]) auditReader.createQuery()
+                .forRevisionsOfEntity(expiredUser.getClass(), false, true)
+                .add(AuditEntity.id().eq(expiredUser.getId()))
+                .add(AuditEntity.revisionNumber().minimize()
+                        .computeAggregationInInstanceContext())
+                .getSingleResult();
+        CustomRevisionEntity first = (CustomRevisionEntity) firstRevision[1];
+        // Update the timestamp of the revision so it appears to have been created 5 days ago
+        ZonedDateTime expDateTime = ZonedDateTime.now().minus(Period.ofDays(5));
+        first.setTimestamp(Date.from(expDateTime.toInstant()));
+        EntityManager entityManager = ((EntityManager) ReflectionTestUtils
+                .getField(revisionService, "entityManager"));
+        entityManager.persist(first);
+
+        // make sure when we reload the expired user we have the new created date
+        assertThat(revisionService.getAuditInfo(expiredUser).getCreatedAt()).isEqualTo(expDateTime);
+
+        // Now we know we have an 'old' user in the database, we can test our deletion method
+        int numSubjects = subjectRepository.findAll().size();
+        subjectService.removeNotActivatedSubjects();
+        List<Subject> subjects = subjectRepository.findAll();
+        // make sure have actually deleted some users, otherwise this test is pointless
+        assertThat(numSubjects - subjects.size()).isEqualTo(1);
+        // remaining users should be either activated or have a created date less then 3 days ago
+        ZonedDateTime cutoff = ZonedDateTime.now().minus(Period.ofDays(3));
+        subjects.forEach(s -> assertThat(s.getUser().getActivated() || revisionService
+                .getAuditInfo(s).getCreatedAt().isAfter(cutoff)).isTrue());
+        commitTransactionAndStartNew();
     }
 }

--- a/src/test/java/org/radarcns/management/service/UserServiceIntTest.java
+++ b/src/test/java/org/radarcns/management/service/UserServiceIntTest.java
@@ -8,11 +8,9 @@ import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.radarcns.auth.config.Constants;
 import org.radarcns.management.ManagementPortalTestApp;
-import org.radarcns.management.domain.Subject;
 import org.radarcns.management.domain.User;
 import org.radarcns.management.domain.audit.CustomRevisionEntity;
 import org.radarcns.management.repository.CustomRevisionEntityRepository;
-import org.radarcns.management.repository.SubjectRepository;
 import org.radarcns.management.repository.UserRepository;
 import org.radarcns.management.repository.filters.UserFilter;
 import org.radarcns.management.service.dto.UserDTO;
@@ -82,9 +80,6 @@ public class UserServiceIntTest {
 
     @Autowired
     private EntityManagerFactory entityManagerFactory;
-
-    @Autowired
-    private SubjectRepository subjectRepository;
 
     private UserDTO userDto;
 
@@ -193,7 +188,7 @@ public class UserServiceIntTest {
 
     @Test
     public void testFindNotActivatedUsersByCreationDateBefore() {
-        User expiredUser = addExpiredUser();
+        User expiredUser = addExpiredUser(userRepository);
         commitTransactionAndStartNew();
 
         AuditReader auditReader = ((AuditReader) ReflectionTestUtils
@@ -230,46 +225,6 @@ public class UserServiceIntTest {
     }
 
     @Test
-    public void testFindNotActivatedSubjectsByCreationDateBefore() {
-        User expiredUser = addExpiredUser();
-        Subject expiredSubject = new Subject();
-        expiredSubject.setUser(expiredUser);
-        subjectRepository.save(expiredSubject);
-        commitTransactionAndStartNew();
-
-        AuditReader auditReader = ((AuditReader) ReflectionTestUtils
-                .getField(revisionService, "auditReader"));
-        Object[] firstRevision = (Object[]) auditReader.createQuery()
-                .forRevisionsOfEntity(expiredUser.getClass(), false, true)
-                .add(AuditEntity.id().eq(expiredUser.getId()))
-                .add(AuditEntity.revisionNumber().minimize()
-                        .computeAggregationInInstanceContext())
-                .getSingleResult();
-        CustomRevisionEntity first = (CustomRevisionEntity) firstRevision[1];
-        // Update the timestamp of the revision so it appears to have been created 5 days ago
-        ZonedDateTime expDateTime = ZonedDateTime.now().minus(Period.ofDays(5));
-        first.setTimestamp(Date.from(expDateTime.toInstant()));
-        EntityManager entityManager = ((EntityManager) ReflectionTestUtils
-                .getField(revisionService, "entityManager"));
-        entityManager.persist(first);
-
-        // make sure when we reload the expired user we have the new created date
-        assertThat(revisionService.getAuditInfo(expiredUser).getCreatedAt()).isEqualTo(expDateTime);
-
-        // Now we know we have an 'old' user in the database, we can test our deletion method
-        int numUsers = userRepository.findAll().size();
-        userService.removeNotActivatedUsers();
-        List<User> users = userRepository.findAll();
-        // make sure have actually deleted some users, otherwise this test is pointless
-        assertThat(numUsers - users.size()).isEqualTo(1);
-        // remaining users should be either activated or have a created date less then 3 days ago
-        ZonedDateTime cutoff = ZonedDateTime.now().minus(Period.ofDays(3));
-        users.forEach(u -> assertThat(u.getActivated() || revisionService.getAuditInfo(u)
-                .getCreatedAt().isAfter(cutoff)).isTrue());
-        commitTransactionAndStartNew();
-    }
-
-    @Test
     public void assertThatAnonymousUserIsNotGet() {
         final PageRequest pageable = new PageRequest(0, (int) userRepository.count());
         final Page<UserDTO> allManagedUsers = userService.findUsers(new UserFilter(), pageable);
@@ -278,7 +233,12 @@ public class UserServiceIntTest {
                 .isTrue();
     }
 
-    private User addExpiredUser() {
+    /**
+     * Create an expired user, save it and return the saved object.
+     * @param userRepository The UserRepository that will be used to save the object
+     * @return the saved object
+     */
+    public static User addExpiredUser(UserRepository userRepository) {
         User user = new User();
         user.setLogin("expired");
         user.setEmail("expired@expired");

--- a/src/test/java/org/radarcns/management/service/UserServiceIntTest.java
+++ b/src/test/java/org/radarcns/management/service/UserServiceIntTest.java
@@ -1,14 +1,18 @@
 package org.radarcns.management.service;
 
 import org.apache.commons.lang3.RandomStringUtils;
+import org.hibernate.envers.AuditReader;
+import org.hibernate.envers.query.AuditEntity;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.radarcns.auth.config.Constants;
 import org.radarcns.management.ManagementPortalTestApp;
+import org.radarcns.management.domain.Subject;
 import org.radarcns.management.domain.User;
 import org.radarcns.management.domain.audit.CustomRevisionEntity;
 import org.radarcns.management.repository.CustomRevisionEntityRepository;
+import org.radarcns.management.repository.SubjectRepository;
 import org.radarcns.management.repository.UserRepository;
 import org.radarcns.management.repository.filters.UserFilter;
 import org.radarcns.management.service.dto.UserDTO;
@@ -19,11 +23,15 @@ import org.springframework.boot.test.context.SpringBootTest;
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.PageRequest;
 import org.springframework.test.context.junit4.SpringRunner;
+import org.springframework.test.util.ReflectionTestUtils;
 import org.springframework.transaction.annotation.Transactional;
 
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
 import java.time.Period;
 import java.time.ZonedDateTime;
 import java.util.Date;
+import java.util.List;
 import java.util.Optional;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -67,13 +75,27 @@ public class UserServiceIntTest {
     private UserService userService;
 
     @Autowired
+    private RevisionService revisionService;
+
+    @Autowired
     private CustomRevisionEntityRepository revisionEntityRepository;
+
+    @Autowired
+    private EntityManagerFactory entityManagerFactory;
+
+    @Autowired
+    private SubjectRepository subjectRepository;
 
     private UserDTO userDto;
 
     @Before
     public void setUp() {
         userDto = userMapper.userToUserDTO(createEntity());
+        ReflectionTestUtils.setField(revisionService, "revisionEntityRepository",
+                revisionEntityRepository);
+        ReflectionTestUtils.setField(revisionService, "entityManagerFactory", entityManagerFactory);
+        ReflectionTestUtils.setField(userService, "userMapper", userMapper);
+        ReflectionTestUtils.setField(userService, "userRepository", userRepository);
     }
 
     /**
@@ -174,22 +196,21 @@ public class UserServiceIntTest {
         User expiredUser = addExpiredUser();
         commitTransactionAndStartNew();
 
-        // Get the first revision of our new user (there should only be one anyway)
-        CustomRevisionEntity firstRevision = userRepository.findRevisions(expiredUser.getId())
-                .getContent().get(0).getMetadata().getDelegate();
-
+        AuditReader auditReader = ((AuditReader) ReflectionTestUtils
+                .getField(revisionService, "auditReader"));
+        Object[] firstRevision = (Object[]) auditReader.createQuery()
+                .forRevisionsOfEntity(expiredUser.getClass(), false, true)
+                .add(AuditEntity.id().eq(expiredUser.getId()))
+                .add(AuditEntity.revisionNumber().minimize()
+                        .computeAggregationInInstanceContext())
+                .getSingleResult();
+        CustomRevisionEntity first = (CustomRevisionEntity) firstRevision[1];
         // Update the timestamp of the revision so it appears to have been created 5 days ago
         ZonedDateTime expDateTime = ZonedDateTime.now().minus(Period.ofDays(5));
-        firstRevision.setTimestamp(Date.from(expDateTime.toInstant()));
-        revisionEntityRepository.save(firstRevision);
-
-        /*
-        // need to investigate this, the revisionService does not pick up the saved revision
-        // entity, even with the flush here. Possibly because it has a different entitymanager.
-        // Everything seems fine in normal application lifecycle though, since we're not modifying
-        // existing revision entities there.
-        revisionEntityRepository.flush();
-
+        first.setTimestamp(Date.from(expDateTime.toInstant()));
+        EntityManager entityManager = ((EntityManager) ReflectionTestUtils
+                .getField(revisionService, "entityManager"));
+        entityManager.persist(first);
 
         // make sure when we reload the expired user we have the new created date
         assertThat(revisionService.getAuditInfo(expiredUser).getCreatedAt()).isEqualTo(expDateTime);
@@ -203,7 +224,49 @@ public class UserServiceIntTest {
         // remaining users should be either activated or have a created date less then 3 days ago
         ZonedDateTime cutoff = ZonedDateTime.now().minus(Period.ofDays(3));
         users.forEach(u -> assertThat(u.getActivated() || revisionService.getAuditInfo(u)
-                .getCreatedAt().isAfter(cutoff)).isTrue());*/
+                .getCreatedAt().isAfter(cutoff)).isTrue());
+        // commit the deletion, otherwise the deletion will be rolled back
+        commitTransactionAndStartNew();
+    }
+
+    @Test
+    public void testFindNotActivatedSubjectsByCreationDateBefore() {
+        User expiredUser = addExpiredUser();
+        Subject expiredSubject = new Subject();
+        expiredSubject.setUser(expiredUser);
+        subjectRepository.save(expiredSubject);
+        commitTransactionAndStartNew();
+
+        AuditReader auditReader = ((AuditReader) ReflectionTestUtils
+                .getField(revisionService, "auditReader"));
+        Object[] firstRevision = (Object[]) auditReader.createQuery()
+                .forRevisionsOfEntity(expiredUser.getClass(), false, true)
+                .add(AuditEntity.id().eq(expiredUser.getId()))
+                .add(AuditEntity.revisionNumber().minimize()
+                        .computeAggregationInInstanceContext())
+                .getSingleResult();
+        CustomRevisionEntity first = (CustomRevisionEntity) firstRevision[1];
+        // Update the timestamp of the revision so it appears to have been created 5 days ago
+        ZonedDateTime expDateTime = ZonedDateTime.now().minus(Period.ofDays(5));
+        first.setTimestamp(Date.from(expDateTime.toInstant()));
+        EntityManager entityManager = ((EntityManager) ReflectionTestUtils
+                .getField(revisionService, "entityManager"));
+        entityManager.persist(first);
+
+        // make sure when we reload the expired user we have the new created date
+        assertThat(revisionService.getAuditInfo(expiredUser).getCreatedAt()).isEqualTo(expDateTime);
+
+        // Now we know we have an 'old' user in the database, we can test our deletion method
+        int numUsers = userRepository.findAll().size();
+        userService.removeNotActivatedUsers();
+        List<User> users = userRepository.findAll();
+        // make sure have actually deleted some users, otherwise this test is pointless
+        assertThat(numUsers - users.size()).isEqualTo(1);
+        // remaining users should be either activated or have a created date less then 3 days ago
+        ZonedDateTime cutoff = ZonedDateTime.now().minus(Period.ofDays(3));
+        users.forEach(u -> assertThat(u.getActivated() || revisionService.getAuditInfo(u)
+                .getCreatedAt().isAfter(cutoff)).isTrue());
+        commitTransactionAndStartNew();
     }
 
     @Test

